---
title: "C. R and Bioconductor for Single Cell Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{C. R and Bioconductor for Single Cell Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# To use this workshop

1. Visit https://workshop.bioconductor.org/
2. Register or log-in
3. choose the 'ARTNet 2023'
4. Wait a minute or so
5. Click to open RStudio
6. In RStudio, choose 'File' / 'Open File...' / 'vignettes/a_r.Rmd'

    ![](images/galaxy-screen.png)

# Introduction

There are rich, well-documented workflows for single-cell analysis in
_R_.  We'll point to these resources and outline overall workflows,
but will only perform illustrative calculations. Some steps require
both computing time and resources that aren't suitable for workshop
presentation so we focus on a 'classic' data set, the 10x Genomics
Peripheral Blood Mononuclear Cells (PBMC) dataset conssists of 2,700
cells. Analysis of data sets with say 50,000 cells might require 32-
or 64 GB of memory and take up to an hour for an end-to-end analysis;
one way to access a computer with the much memory might be through a
commercial (Google, Azure, AWS) cloud, or on a research network like
the [AnVIL][] computational cloud.

[AnVIL]: https://anvilproject.org

# Seurat

Based on [Getting Started with Seurat][] and the [Seurat - Guided
Clustering Tutorial][Seurat tutorial]. The guided tutorial covers

- Data input
- Quality Control
- Normalization
- Identification of highly variable features
- Scaling the data
- Linear dimensional reduction
- Clusters
- Non-linear dimensional reduction (e.g., UMAP)
- Differential expression
- Cell type assignment

To illustrating working with `Seurat`, we take the first few steps
through this workflow.

## Workflow

Here we re-capitulate some of the initial steps in the [Seurat
tutorial][].

Data retrieval. Here we retrieve the raw data to a local cache. It is
distributed as a compressed 'tar' archive, so we `untar()` it to a
temporary location.

```{r}
## download to local cache
url <- paste0(
    "https://cf.10xgenomics.com/samples/cell/pbmc3k/",
    "pbmc3k_filtered_gene_bc_matrices.tar.gz"
)
pbmc3k_tar_gz <- BiocFileCache::bfcrpath(rname = url)

## 'untar' to a temporary location
pbmc3k_directory <- tempfile()
untar(pbmc3k_tar_gz, exdir = pbmc3k_directory)

## location of the relevant data
pbmc3k_data_dir <-
    file.path(pbmc3k_directory, "filtered_gene_bc_matrices", "hg19")
```

Following the [Seurat tutorial][], we read the data into a sparse
matrix, and then into a Seurat object.

```{r}
library(Seurat)

## read the 10x data as a sparse matrix
pbmc.data <- Read10X(data.dir = pbmc3k_data_dir)

class(pbmc.data)

dim(pbmc.data) # 32738 features (genes) x 2700 samples (cells)

## create a Seurat object from the count matrix
pbmc <- CreateSeuratObject(
    counts = pbmc.data,
    project = "pbmc3k",
    ## filter some features (present in less than 3 cells) and cells
    ## (fewer than 200 features)
    min.cells = 3, min.features = 200
)

pbmc # 13714 features x 2700 samples

pbmc[["RNA"]]
```

The first step in the workflow is to perform quality control. Start by
identifying the mitochondrial genes. In this particular example,
mitchondrial genes are annotated based on the feature name.

```{r}
## The [[ operator can add columns to object metadata. This is a great
## place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

Visualize some metrics...

```{r}
VlnPlot(
    pbmc,
    features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
    ncol = 3
)
```

...and use *ad hoc* criteria to filter features; `subset()` is a base
*R* function.

```{r}
pbmc <- subset(
    pbmc,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5
)

pbmc
```

The workflow continues from here...

# Bioconductor

The following is based on [Orchestrating Single-Cell Analysis with
Bioconductor][OSCA]. This resource is separated into introductory,
basic, advanced, and multi-sample sections, with a collection of
workflows illustrating use. A good place to start is with a workflow
to get a feel for data analysis, and then to refer back to earlier
sections for more detailed operations or understanding.

We'll follow the [Unfiltered human PBMCs][] workflow for another PBMC
data, but use the same dataset as in the Seurat section earlier in
this article.

[OSCA]: http://bioconductor.org/books/release/OSCA.basic/
[Unfiltered human PBMCs]: https://bioconductor.org/books/release/OSCA.workflows/unfiltered-human-pbmcs-10x-genomics.html

## Workflow

Data retrieval is a outlined above. After `untar()`ing the data
`pbmc3k_data_dir` is the location to use for data input.

```{r, message = FALSE}
library(DropletUtils)
library(scuttle)
```

```{r}
pbmc3k <- read10xCounts(pbmc3k_data_dir, col.names = TRUE)

pbmc3k
```

Skipping to the quality control step, let's identify mitochondrial
genes. In this data set, we can identify the mitochondrial genes by
looking at the `Symbol` column of `rowData()`

```{r}
rowData(pbmc3k) |>
    dplyr::as_tibble() |>
    dplyr::filter(startsWith(Symbol, "MT-"))

is_mitochondrial <-
    rowData(pbmc3k)$Symbol |>
    startsWith("MT-") |>
    which()
```

Calculate per-cell quality control metrics...

```{r}
stats <- perCellQCMetrics(pbmc3k, subsets=list(Mito=is_mitochondrial))
stats |> dplyr::as_tibble()

high_mito <- isOutlier(stats$subsets_Mito_percent, type="higher")
table(high_mito)
```

...and use `high_mito` to create a subset of the results

```{r}
pbmc3k[, !high_mito]
```

The workflow continues...!

# Brief Commentary

It is worthwhile to compare the *Seurat* and *SingleCellExperiment*
work flows.

Actually, a first question is 'why not just have a single function
that takes the analysis from begining to end?' Answers might be

1. It is educational to understand the data transformations and their
   assummptions required in an analysis.

2. No two data sets are the same, so at any step some manipulation
   (e.g., to identify mitochondrial genes), choice of algorithm, or
   interpretation of results might be required.

What about the difference between the *Seurat* and *Bioconductor*
workflows? In broad detail the analyses accomplish the same
ends. *Seurat* is more standardized and widely used, while
*Bioconductor* may be less standardized and more flexible.

[Getting Started with Seurat]: https://satijalab.org/seurat
[Seurat tutorial]: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

# Summary

## Session information

This document was produced with the following *R* software:

```{r session_info}
sessionInfo()
```

```{r include = FALSE}
## clean up .GlobalEnv at end of vignette
rm(list = ls(envir = .GlobalEnv, all.names = TRUE), envir = .GlobalEnv)
```
